x-pack/plugins/monitoring/server/kibana_monitoring/index.js

```diff
@@ -4,6 +4,70 @@
  * you may not use this file except in compliance with the Elastic License.
  */
 
-export { createCollectorSet } from './create_collector_set';
-export { getKibanaUsageCollector } from './collectors/get_kibana_usage_collector';
-export { UsageCollector } from './classes/usage_collector';
+import { LOGGING_TAG, KIBANA_MONITORING_LOGGING_TAG, } from '../../common/constants';
+import { CollectorSet } from './classes';
+import {
+  getKibanaUsageCollector,
+  getOpsStatsCollector,
+  getSettingsCollector,
+} from './collectors';
+import { BulkUploader } from './bulk_uploader';
+import { monitoringBulk } from './lib/monitoring_bulk';
+import { sendBulkPayload } from './lib/send_bulk_payload';
+import { getCollectorTypesCombiner } from './lib';
+
+/**
+ * Initialize different types of Kibana Monitoring
+ * TODO: remove this in 7.0
+ * - Ops Events - essentially Kibana's /api/status
+ * - Usage Stats - essentially Kibana's /api/stats
+ * - Kibana Settings - select uiSettings
+ * @param kbnServer {Object} manager of Kibana services - see `src/server/kbn_server` in Kibana core
+ * @param server {Object} HapiJS server instance
+ */
+export function initKibanaMonitoring(kbnServer, server) {
+  const mainXpackInfo = server.plugins.xpack_main.info;
+  const mainMonitoring = mainXpackInfo.feature('monitoring');
+
+  let collectorSet = null;
+  if (mainXpackInfo && mainMonitoring.isAvailable() && mainMonitoring.isEnabled()) {
```

Comment:
This is a dangerous race condition. If Kibana starts up before/while Elasticsearch is starting (or happens to be non-responsive for some reason during this time period), then Kibana monitoring is guaranteed to be disabled until Kibana is restarted. Instead, I think that we would be better off stopping/starting the collector based on this logic (or, continuing to blindly start it). Since we can listen to the `xpack_main.info` object, we should be safe to do that.

Also, by not returning the `collectorSet`, we force other plugins to validate the `CollectorSet`'s existence before using it in addition to the monitoring plugin being enabled.

Reply:
In the next version of this code, `collectorSet` will be moved to `src/server/usage` in core Kibana. There, it will be initiated before any of the plugins and core plugins are. Plugins will access it like:
```
const { collectorSet } = server.usage;
```

This part of the code looks like a miss in the refactoring - the CollectorSet constructor only needs `server` passed in, and it needs that for logging.

Reply:
`collectorSet` doesn't need to have conditional instantiation/startup, but the bulk uploader's start method probably does